dhcp: don't enforce hardware address length for sd_dhcp_client_set_client_id()

thom311 opened the pull request:
sd_dhcp_client_set_client_id() is the only API for setting a raw client-id.
All other setters are more restricted and only allow to set a type 255 DUID.

Also, dhcp4_set_client_identifier() is only one caller, which already
does:

                r = sd_dhcp_client_set_client_id(link->dhcp_client,
                                                 ARPHRD_ETHER,
                                                 (const uint8_t *) &link->mac,
                                                 sizeof(link->mac));

and hence ensures that the data length is indeed ETH_ALEN.

Drop additional input validation from sd_dhcp_client_set_client_id(). The client-id
is an opaque blob, and if a caller wants to set type 1 (ethernet) or type 32
(infiniband) with unexpected address length, it should be allowed. The actual
client-id is not relevant to the DHCP client, and it's the responsibility of the
caller to ensure that the client-id is set accordingly.

For example, in NetworkManager you can configure all the bytes of the
client-id, including such "invalid" settings. I think that makes sense,
if the user wants to configure it, it should just allow it. Anyway, even
if such configuration would be prevented, it would be the responsibility
of the higher layers to reject such client-ids, not not fail later when
tring to use it in sd_dhcp_client_set_client_id().

thom311 commented:
btw, another reason why this is wrong, is because for ARPHRD_INFINIBAND we should only use the last 8 bytes, not the full 20 bytes. This is at least what dhclient does.

-- not that it matters for networkd, since that will call `sd_dhcp_client_set_client_id()` only for ethernet.

poettering commented:
hmm, could you add a debug log msg here at least in case the sizes are weird? that way you won't have to see this normally, but if somebody wants to debug something they can find it?

thom311 commented:
done. How about v2? (also extended commit message)

poettering commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -300,27 +300,17 @@ int sd_dhcp_client_set_client_id(
             memcmp(&client->client_id.raw.data, data, data_len) == 0)
                 return 0;
 
+        if ((type == ARPHRD_ETHER      && data_len != ETH_ALEN) ||
```
hmm, we usually don't align things like this in tables, but i guess this doesn't matter

poettering commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -300,27 +300,17 @@ int sd_dhcp_client_set_client_id(
                 return 0;
 
+        if ((type == ARPHRD_ETHER      && data_len != ETH_ALEN) ||
+            (type == ARPHRD_INFINIBAND && data_len != 8))
```
could you add a short comment here why you comapre with 8 here? i.e. the stuff you had in the commit msg?

poettering commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -300,27 +300,17 @@ int sd_dhcp_client_set_client_id(
+        if ((type == ARPHRD_ETHER      && data_len != ETH_ALEN) ||
+            (type == ARPHRD_INFINIBAND && data_len != 8))
+                log_dhcp_client(client, "Changing client ID to hardware type %u with "
+                                "unexpected address length %u",
```
please use `%zu` here, rather than %u, so that there's no need to cast.

poettering commented on src/libsystemd-network/sd-dhcp-client.c:
```diff
@@ -300,27 +300,17 @@ int sd_dhcp_client_set_client_id(
+            (type == ARPHRD_INFINIBAND && data_len != 8))
+                log_dhcp_client(client, "Changing client ID to hardware type %u with "
+                                "unexpected address length %u",
+                                (unsigned) type, (unsigned) data_len);
```
the first cast is unnecessary, as anything smaller than int/unsigned int is upcast to int/unsigned int anyway in varargs arguments...

thom311 commented:
v3, I addressed the comments so far.

What's new:

- a new commit "dhcp: fix sd_dhcp_client_set_client_id() for infiniband addresses" that changes the behaviour regarding `INFINIBAND_ALEN`. The point is to have an individual commit and (somewhat) of a reasoning in the commit message.
- new commit "dhcp6: don't enforce DUID content for sd_dhcp6_client_set_duid()", which does something similar for DHCPv6.
